src/Pinto/GenServer.purs

```diff
@@ -0,0 +1,215 @@
+-- Public types
+--------------------------------------------------------------------------------
+data CallResult reply cont stop state
+  = CallReply reply state
+  | CallReplyWithTimeout reply state Int
+  | CallReplyHibernate reply state
+  | CallReplyContinue reply state cont
+  | CallNoReply state
+  | CallNoReplyWithTimeout state Int
+  | CallNoReplyHibernate state
+  | CallNoReplyContinue state cont
+  | CallStopReply stop reply state
+  | CallStopNoReply stop state
+
+instance mapCallResult :: Functor (CallResult reply cont stop) where
+  map f (CallReply reply state) = (CallReply reply (f state))
+  map f (CallReplyWithTimeout reply state n) = CallReplyWithTimeout reply (f state) n
+  map f (CallReplyHibernate reply state) = CallReplyHibernate reply (f state)
+  map f (CallReplyContinue reply state cont) = CallReplyContinue reply (f state) cont
+  map f (CallNoReply state) = CallNoReply (f state)
+  map f (CallNoReplyWithTimeout state n) = CallNoReplyWithTimeout (f state) n
+  map f (CallNoReplyHibernate state) = CallNoReplyHibernate (f state)
+  map f (CallNoReplyContinue state cont) = CallNoReplyContinue (f state) cont
+  map f (CallStopReply stop reply state) = CallStopReply stop reply (f state)
+  map f (CallStopNoReply stop state) = CallStopNoReply stop (f state)
+
+
+data CastResult state
+  = NoReply state
+  -- | NoReplyTimeout state Int
+  -- | NoReplyHibernate state
+  -- | NoReplyContinue state cont
+  -- | Stop stop state
+
+instance mapCastResult :: Functor CastResult where
+  map f (NoReply state) = NoReply (f state)
+
+
+type InfoResult state = CastResult state
+type ContinueResult state = CastResult state
+
+type ResultT result cont stop msg state = ReaderT (Context cont stop msg state) Effect result
+
+
+-- TODO make order of type variables consistent
+type InitFn cont stop msg state = ResultT (InitResult cont state) cont stop msg state
+type CallFn reply cont stop msg state = state -> ResultT (CallResult reply cont stop state) cont stop msg state
+type CastFn cont stop msg state = state -> ResultT (CastResult state) cont stop msg state
+type ContinueFn cont stop msg state = cont -> state -> ResultT (ContinueResult state) cont stop msg state
+type InfoFn cont stop msg state = msg -> state -> ResultT (InfoResult state) cont stop msg state
+
+-- -- | Type of the callback invoked during a gen_server:handle_cast
+-- type Cast state msg = ResultT (CastResult state) state msg
+
+
+data ServerRunning cont state
+  = InitOk state
+  | InitOkTimeout state Int
+  | InitOkContinue state cont
+  | InitOkHibernate state
+
+data ServerNotRunning
+  = InitStop Foreign
+  | InitIgnore
+
+type InitResult cont state = Either ServerNotRunning (ServerRunning cont state )
+
+-- Can't do a functor instance over a type synonym, so just have a function instead
+mapInitResult :: forall state state' cont. (state -> state') -> InitResult cont state -> InitResult cont state'
```

Comment:
I can understand wanting to capture "success" vs "failure" in the init results, but the introduction of Either on top of an ADT doesn't really add anything unless you're using 'either' functionality somewhere *on* that result - @nwolverson ?

Reply:
Yep that looks odd - if the ADT is representing init results it seems like it's just one thing, and stick a function `:: InitResult -> Either _ state` if required on top of that